### ISSUE TITLE
Some commands fail on YAML that only contains a list

### ISSUE THREAD

**Issue by bzub** [GoogleContainerTools/kpt#1218]
This produces an error: `wrong Node Kind for  expected: MappingNode was SequenceNode`. Seems like `SequenceNode` should be allowed?

`FieldClearer` and `FieldMatcher` types seem to disallow this.

```sh
(
REPRO_HOME=$(mktemp -d)

cat <<EOF >$REPRO_HOME/list_only.yaml
# A comment
- key: value
- string
EOF

kpt pkg init $REPRO_HOME
kpt cfg create-setter $REPRO_HOME key value --stack-trace
)
```

```sh
echo '- key: value\n- string'|kpt cfg cat
error: wrong Node Kind for  expected: MappingNode was SequenceNode: value: {- key: value
- string}
```

**Comment by Bobgy** [comment]
I'm seeing the same error when `kpt pkg get` a package that contains a yaml file which is essentially a list (it was used as a kustomize json patch).
I could workaround the issue by using `kpt pkg get --auto-set=false`, but `kpt cfg set -R` still raise the same exception.

**Comment by mikebz** [comment]
@natasha41575 @runewake2 I'd consider picking up this ticket if you are blocked on one of the bigger tasks.

**Comment by natasha41575** [comment]
@bzub Can you describe a use case for having a yaml file that is only a list, and what you expect the output to be? For now kpt parses all yaml files as kubernetes resources which is why the error is occurring. 

One quick alternative to get around this error is to rename the file so that it doesn't have a yaml extension (this way, kpt will ignore it). In the future there may be a way to add files to an ignore file to prevent issues like this.

**Comment by seh** [comment]
Such top-level lists arise in JSON Patch files written as YAML. It's necessary to place such patch definitions in files separate from a _Krmfile_ or _Kptfile_ file in order to use setters and substitutions in the content of the patch.

See #510 and  #1324 for more of that story.

**Comment by phanimarupaka** [comment]
@bzub kpt can only parse and process valid Kubernetes Resource files. Since the file has .yaml extension and is not a valid kubernetes resource, you must intimate kpt to not parse it by creating a file with name .krmignore in the package directory and add the file name`(list_only.yaml)` to it. This way, kpt ignores the file and leave it untouched.

@Bobgy The remote package in github should have .krmignore file as mentioned above so that when you pull the package onto local the file is also fetched but will not obstruct the regular flow.

@seh Non-KRM files can only be ignored by kpt as mentioned above and you can't add setter references to such files. 

cc @mikebz @natasha41575 @etefera @Shell32-Natsu 
We should have an issue to have the clear error message that kpt doesn't parse non-KRM resources and should be added to .krmignore file.

**Comment by seh** [comment]
I'm not using _kpt_ here, though; I'm using _kustomize_, and it suffers from the same problem. Why would you assume that all YAML files are Kubernetes resources if they're not mentioned as values in the "resources" field of a _kustomization.yaml_ file?

**Comment by seh** [comment]
Also, we should be able to interpolate patch files; there's no other way to use a setter- or substitution-injected value as a key in a mapping entry. This comes up when building up annotations and labels whose keys are predicated on some injected value, though not always in "metadata.labels" or "metadata.annotations" fields.

**Comment by phanimarupaka** [comment]
@seh I would say it is a limitation with kpt/kustomize the way we parse and track the yaml file nodes. This is more of implementation detail but trying to explain here. 

When kpt/kustomize reads the the files to process, it adds path(file path) and index(index in case there are multiple yaml nodes in a file) annotations to those yaml nodes in order to know where to write them back. If there is only a sequence node in the the file, then it would not be possible for kpt/kustomize to add such annotations and thus would loose the file context. So kpt/kustomize expects the top level node to be a Mapping node for it to add such annotations and if it is not, you see this error. Kpt/kustomize has to fundamentally change the way files are parsed and tracked internally to solve this issue.

**Comment by seh** [comment]
Thank you for the explanation.

Not knowing the trouble you've been through, has the team considered an approach of storing such a top-level list internally as a mapping, with a sentinel type or annotation that says, "This is a placeholder, where the file content is the value of the 'data' field," or something to that effect? That is, during file output, you wouldn't write the whole object in that case, but would instead only write the value of one of its fields.

**Comment by phanimarupaka** [comment]
We have considered that approach of storing such metadata separate from actual nodes in memory, but it would not work well if the output is piped though chain of kpt commands(e.g. kpt fn source | kpt fn run| kpt fn sink). Adding such information to yaml node itself has been more robust solution so far. 

I would recommend you to have one of the work arounds to solve this problem in your case, may be wrap that list inside a mapping node and update your logic. Or any other ideas from you are welcome.

**Comment by seh** [comment]
If we were able to apply setters and substitutions _inside of string values_, I'd be satisfied. Here's where it comes up: Say that inside of my _kustomization.yaml_ file, I have a "patches" field entry, and I want to use an inline patch definition. That's a string mapping entry value:
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
patches:
- target:
    kind: Something
  patch: |-
    - op: add
      path: placeholder # {"$openapi":"some-json-pointer"}
      value: known
```
If I could annotate that patch somehow to tell _kustomize_ or _kpt_ to parse that string as YAML,  interpolate it with setters and substitutions, and write the transformed YAML document back to that value as a string, I'd be set.

As it stands, because the value of the "patches.patch" field is a string, _kustomize_ doesn't look inside of it for potential setter or substitution annotations.

**Comment by Bobgy** [comment]
Thanks! .krmignore will be enough for my current use-case

**Comment by phanimarupaka** [comment]
@seh So summarizing your request...

If you start with 

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
patches:
- target:
    kind: Something
  patch: |-
    - op: add
      path: placeholder
      value: known
```

and do

`kpt cfg create-setter . some-json-pointer placeholder`

The output should be

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
patches:
- target:
    kind: Something
  patch: |-
    - op: add
      path: placeholder # {"$kpt-set":"some-json-pointer"}
      value: known
```

and later on

`kpt cfg set . some-json-pointer new-placeholder`

should give output 

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
patches:
- target:
    kind: Something
  patch: |-
    - op: add
      path: new-placeholder # {"$kpt-set":"some-json-pointer"}
      value: known
```

**Comment by seh** [comment]
Yes, you have understood me perfectly.

We would need some way to indicate that that "patches[0].patch" value is not "just a string," but rather a YAML document that's in need of interpolation.

By the way, I forgot to mention it yesterday, but people may also have YAML documents (including those with a top-level sequence) that are not Kubernetes manifests as input to things like the _ConfigMap_ and _Secret_ generators. Needing to ignore those—and them being excluded from setter and substitution interpolation—is unfortunate.

**Comment by phanimarupaka** [comment]
@seh This should work as you expected but it is not when I just tried. So it must be a bug with parsing and is not related this parent issue. I created another issue for better tracking and will work on addressing that. https://github.com/GoogleContainerTools/kpt/issues/1342

**Comment by phanimarupaka** [comment]
@bzub Does .krmignore also solves your use-case? Can we close this issue if it does?

**Comment by phanimarupaka** [comment]
@dzub closing this issue. Please feel free to reopen if you need further help.

**Comment by Bobgy** [comment]
Hi @phanimarupaka, I tried using `.krmignore`, but I kept getting the `wrong Node Kind error`.
Can you clarify when that feature was supported?
Is it in `kpt 0.37.1`?
And what is its syntax?

I'm getting this error message:
```
kpt cfg set -R . xxxxx
error: /<redacted>/upstream/manifests/namespace-install/overlays/argo-server-deployment.yaml: wrong Node Kind for  expected: MappingNode was SequenceNode: value: {- op: add
  path: /spec/template/spec/containers/0/args/-
  value: --namespaced}
```

I put a `.krmignore` file in `upstream/manifests` with the following content:

```
namespace-install/overlays/argo-server-deployment.yaml
argo-server-deployment.yaml
```

However, neither the relative path, nor the file name seem to be read. I'm still getting the error message during kpt cfg set.

**Comment by Bobgy** [comment]
Ohhh, I think I figured out
`.krmignore` can only be used to ignore folders, not files.

Is that correct? Can we have official documentation for it? It's a little hard digging through how exactly it works.

**Comment by Bobgy** [comment]
Hmm, another caveat found, it seems `.krmignore` only takes effect when my working dir directly contains it. This seems make it unsable, I publish packages, if a `.krmignore` in the upstream package doesn't have effect, all my consumers need to set up a `.krmignore` for themselves?

**Comment by phanimarupaka** [comment]
When package consumers pull the package, krmignore file is also pulled onto the package. So package consumers need not set it up. Can you point me to the example repo and steps to reproduce the problem.
